### ISSUE TITLE
bgfx assertion failure when starting firefox

### ISSUE THREAD

**Issue by cuavas** [mamedev/mame#5830]
Using Windows 10 Pro 1803, notebook with Intel UHD 630 (driver 24.20.100.6195) and NVIDIA GeForce MX130 (driver 23.21.13.8908) GPUs.  Also happens on Fedora 28.

I get an assertion failure when running LaserDisc games (e.g. Fire Fox) with any bgfx backend:
```
Assertion failed: effect->uniform(m_sampler) != nullptr, file ../../../../../src/osd/modules/render/bgfx/inputpair.cpp, line 49

Thread 1 hit Breakpoint 1, 0x00007ffe37c2edd7 in msvcrt!abort () from /c/Windows/System32/msvcrt.dll
(gdb) where
#0  0x00007ffe37c2edd7 in msvcrt!abort () from /c/Windows/System32/msvcrt.dll
#1  0x00007ffe37c2d61b in msvcrt!_wassert () from /c/Windows/System32/msvcrt.dll
#2  0x0000000006e7d16e in bgfx_input_pair::bind(bgfx_effect*, int) const ()
#3  0x0000000006ea5e29 in bgfx_chain_entry::submit(int, chain_manager::screen_prim&, texture_manager&, unsigned short, unsigned short, unsigned short, float, float, float, float, unsigned int, bool, unsigned long long, int) ()
#4  0x0000000006ea1c7e in bgfx_chain::process(chain_manager::screen_prim&, int, int, texture_manager&, osd_window&, unsigned long long) ()
#5  0x0000000006e64b01 in chain_manager::process_screen_quad(unsigned int, unsigned int, chain_manager::screen_prim&, osd_window&) ()
#6  0x0000000006e64c02 in chain_manager::process_screen_chains(unsigned int, osd_window&) ()
#7  0x0000000006e35fba in renderer_bgfx::draw(int) ()
#8  0x0000000006e20ebe in sdl_window_info::update() ()
#9  0x0000000006e1ec6f in sdl_osd_interface::update(bool) ()
#10 0x000000000685ddfa in video_manager::frame_update(bool) ()
#11 0x00000000068234a2 in screen_device::vblank_begin() ()
#12 0x000000000682662d in screen_device::device_timer(emu_timer&, unsigned int, int, void*) ()
#13 0x000000000681c1c7 in device_scheduler::timeslice() ()
#14 0x00000000067d8cbc in running_machine::run(bool) ()
#15 0x0000000006be287c in mame_machine_manager::execute() ()
#16 0x0000000006c5e534 in cli_frontend::start_execution(mame_machine_manager*, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&) ()
#17 0x0000000006c5e8c6 in cli_frontend::execute(std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >&) ()
#18 0x0000000006be0a89 in emulator_info::start_frontend(emu_options&, osd_interface&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >&) ()
#19 0x0000000006e1b342 in SDL_main ()
#20 0x0000000006e1a88c in main_getcmdline ()
#21 0x000000000040138c in __tmainCRTStartup () at E:/mingwbuild/mingw-w64-crt-git/src/mingw-w64/mingw-w64-crt/crt/crtexe.c:339
#22 0x00000000004014bb in mainCRTStartup () at E:/mingwbuild/mingw-w64-crt-git/src/mingw-w64/mingw-w64-crt/crt/crtexe.c:223
```

I get an outright crash when trying to use the Vulkan backend.  Strangely this happens after confirming the warning about bad dumps and the system information screen.  It seems it can draw some things OK, but crashes when it wants to draw game video.  It can display the system selection menu fine, too.
```
Thread 5 received signal SIGSEGV, Segmentation fault.
[Switching to Thread 11164.0x5708]
0x000000001ca88295 in vk_optimusGetInstanceProcAddr () from /c/Windows/system32/nvoglv64.dll
(gdb) where
#0  0x000000001ca88295 in vk_optimusGetInstanceProcAddr () from /c/Windows/system32/nvoglv64.dll
#1  0x000000001ca8a9f0 in vk_optimusGetInstanceProcAddr () from /c/Windows/system32/nvoglv64.dll
#2  0x0000000006fcedbe in bgfx::vk::RendererContextVK::submit(bgfx::Frame*, bgfx::ClearQuad&, bgfx::TextVideoMemBlitter&) ()
#3  0x0000000009667c1d in bgfx::Context::renderThread(bx::Thread*, void*) ()
#4  0x00007ffe379d3034 in KERNEL32!BaseThreadInitThunk () from /c/Windows/System32/KERNEL32.DLL
#5  0x00007ffe38143691 in ntdll!RtlUserThreadStart () from /c/Windows/SYSTEM32/ntdll.dll
#6  0x0000000000000000 in ?? ()
Backtrace stopped: previous frame inner to this frame (corrupt stack?)
```

The stack trace doesn't look that helpful.  Perhaps I need to look at a different thread?

edit: The crash when using Vulkan was resolved by obtaining a later NVIDIA driver.  As of 06f99c21a08fe76ae833a5587d19c33ed0a035f1 the assertion failure with Fire Fox remains.

edit2: Updated after confirming the assertion failure also occurs on Linux.

**Comment by belegdol** [comment]
It is a wild guess, but could it be an issue with the fact that LD games use yuv texture format? There were issues with palletted drivers fixed by https://github.com/mamedev/mame/commit/ec4bee1371353b2767a1260637ab5317646970ca.

**Comment by belegdol** [comment]
On Fedora 31 with nvidia blob driver, both opengl and vulkan backends crash. The crash can be prevented by passing -bgfx_screen_chains none at the cost of the colours being off as this disables colourspace conversion.
@MooglyGuy, would you mind having a look? Thanks!

**Comment by belegdol** [comment]
I need to revise my previous comment: I had bgfx_screen_chains set to hlsl in my config, apologies for the noise. With -bgfx_screen_chains default the crash is gone too.
Incidentally, ikaruga crashes with hlsl chain too here, but this might be unrelated. I will do a symbols build and report back.

**Comment by belegdol** [comment]
In all 4 cases (firefox and ikaruga with opengl and vulkan) the backtraces for hlsl look very similar:
```
Thread 1 "mame64" received signal SIGSEGV, Segmentation fault.
0x00000000006d87eb in chain_manager::update_screen_textures (this=0x29d77c0, view=<optimized out>, starting_prim=<optimized out>, 
    window=...) at ../../../../../src/osd/modules/render/bgfx/chainmanager.cpp:515
515		return m_screen_prims.size();
(gdb) bt
#0  0x00000000006d87eb in chain_manager::update_screen_textures (this=0x29d77c0, view=<optimized out>, 
    starting_prim=<optimized out>, window=...) at ../../../../../src/osd/modules/render/bgfx/chainmanager.cpp:515
#1  0x0000000000674f4d in renderer_bgfx::draw (this=0x2d0c7c0, update=<optimized out>) at ../../../../../src/emu/render.h:360
#2  0x000000000064d03a in sdl_window_info::update (this=0x29a3f90) at /usr/include/c++/9/bits/unique_ptr.h:352
#3  sdl_window_info::update (this=0x29a3f90) at ../../../../../src/osd/sdl/window.cpp:560
#4  0x000000000064a6dd in sdl_osd_interface::update (this=0x7fffffffcfa0, skip_redraw=<optimized out>)
    at ../../../../../src/osd/sdl/video.cpp:108
#5  0x0000000001b3919b in video_manager::frame_update (this=0x43599c0, from_debugger=from_debugger@entry=false)
    at ../../../../../src/emu/video.h:54
#6  0x0000000001afaade in screen_device::vblank_begin (this=0x2c66d30) at /usr/include/c++/9/bits/unique_ptr.h:352
#7  0x0000000001afd8b5 in screen_device::device_timer (this=0x2c66d30, timer=..., id=<optimized out>, param=<optimized out>, 
    ptr=<optimized out>) at ../../../../../src/emu/screen.cpp:897
#8  0x0000000001af4942 in device_t::timer_expired (ptr=<optimized out>, param=<optimized out>, id=<optimized out>, timer=..., 
    this=<optimized out>) at ../../../../../src/emu/device.h:609
#9  device_scheduler::execute_timers (this=<optimized out>) at ../../../../../src/emu/schedule.cpp:894
#10 device_scheduler::timeslice (this=this@entry=0x7fffffffc298) at ../../../../../src/emu/schedule.cpp:534
#11 0x0000000001a9e568 in running_machine::run (this=this@entry=0x7fffffff6d20, quiet=quiet@entry=false)
    at ../../../../../src/emu/machine.cpp:367
#12 0x0000000000732a25 in mame_machine_manager::execute (this=this@entry=0x29de220)
    at ../../../../../src/frontend/mame/mame.cpp:261
#13 0x00000000007ce73a in cli_frontend::start_execution (this=0x7fffffffcd90, manager=0x29de220, args=...)
    at ../../../../../src/frontend/mame/clifront.cpp:264
#14 0x00000000007ce91e in cli_frontend::execute (this=this@entry=0x7fffffffcd90, 
    args=std::vector of length 12, capacity 12 = {...}) at ../../../../../src/frontend/mame/clifront.cpp:280
#15 0x0000000000730cd8 in emulator_info::start_frontend (options=..., osd=..., args=std::vector of length 12, capacity 12 = {...})
    at ../../../../../src/frontend/mame/mame.cpp:392
#16 0x00000000004df9cb in main (argc=<optimized out>, argv=<optimized out>) at ../../../../../src/osd/sdl/sdlmain.cpp:217
```

**Comment by belegdol** [comment]
It seems like the hlsl issue might be separate. I have built latest git master under windows 10 and found the following:

renderer |  ikaruga default | ikaruga hlsl | firefox default | firefox hlsl
--- | --- | --- | --- | ---
dx9 | crash | crash | crash | crash
dx11 | crash | crash | crash | crash
dx12 | crash | crash | runs | crash
opengl | runs | crash | crash | crash
vulkan | runs | crash | crash | crash

In summary: hlsl crashes across the board while the default screen chain works on opengl and vulkan for ikaruga and on dx12 for firefox.

**Comment by cuavas** [comment]
@belegdol that stack trace is unrelated to either of the things I'm encountering.  I resolved the crash in the NVIDIA Maxwell driver by replacing the NVIDIA driver from Windows Update (apparently version 389) with version 411.12 obtained directly from NVIDIA.

**Comment by belegdol** [comment]
I just tested 9efc500cc00546a78e918915eedf085231e6397a on linux. While there are no crashes with firefox with neither opengl or vulkan, hlsl enabled or not, LD video is not displayed properly with opengl:
![Zrzut ekranu z 2019-11-20 20-53-13](https://user-images.githubusercontent.com/1208329/69273357-7f038500-0bd8-11ea-8e52-5c52b441e19e.png)
with vulkan everything works fine:
![Zrzut ekranu z 2019-11-20 20-53-43](https://user-images.githubusercontent.com/1208329/69273379-8c207400-0bd8-11ea-9c7c-627c168eabad.png)
This is with Fedora 31 with 440.31 nvidia drivers.
Can it be that the shaders are causing some sort of edge case only the nvidia drivers know how to deal with without crashing?

**Comment by cuavas** [comment]
Did you use a debug build (i.e. add `DEBUG=1` to your make flags to enable assertions)?  You won't see an assertion failure if you're not using a debug build.  An assertion failure means something is breaking assumptions, and either the code causing the assumptions to break needs to be fixed, or the assumptions are wrong and the code making the assumptions needs to be fixed.

**Comment by belegdol** [comment]
You are correct, my apologies. I did not know the difference between assertion failure and a crash. With DEBUG=1 build there are indeed assertion failures with both opengl and vulkan:
```
$ ./mame64d -rompath /mnt/openmediavault/emu/mame/roms/ -nowindow -video bgfx -bgfx_backend opengl -bgfx_debug firefox
mame64d: ../../../../../src/osd/modules/render/bgfx/inputpair.cpp:49: void bgfx_input_pair::bind(bgfx_effect*, int32_t) const: Assertion `effect->uniform(m_sampler) != nullptr' failed.
Przerwane (zrzut pamięci)
$ ./mame64d -rompath /mnt/openmediavault/emu/mame/roms/ -nowindow -video bgfx -bgfx_backend vulkan -bgfx_debug firefox
mame64d: ../../../../../src/osd/modules/render/bgfx/inputpair.cpp:49: void bgfx_input_pair::bind(bgfx_effect*, int32_t) const: Assertion `effect->uniform(m_sampler) != nullptr' failed.
```

**Comment by MooglyGuy** [comment]
This assertion failure should now be fixed @cuavas , please close once verified fixed.

**Comment by cuavas** [comment]
Assert fail is fixed, "four squares" with OpenGL appears to be a separate issue.
